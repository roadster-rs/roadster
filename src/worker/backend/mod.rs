use crate::config::service::worker::QueueConfig;
use itertools::Itertools;
use std::collections::{BTreeMap, BTreeSet};

#[cfg(feature = "worker-pg")]
pub mod pg;
#[cfg(feature = "worker-sidekiq")]
pub mod sidekiq;

fn shared_queues<'a>(
    config_queues: &'a Option<BTreeSet<String>>,
    all_queues: &'a BTreeSet<String>,
    dedicated_queues: &'a BTreeMap<String, QueueConfig>,
) -> impl Iterator<Item = &'a String> {
    config_queues
        .as_ref()
        .unwrap_or(all_queues)
        .iter()
        .filter(|queue| !dedicated_queues.contains_key(*queue))
}
