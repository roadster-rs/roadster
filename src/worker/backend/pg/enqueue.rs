use crate::app::context::AppContext;
use crate::config::AppConfig;
use crate::error::RoadsterResult;
use crate::worker::{Enqueuer, Worker, enqueue};
use async_trait::async_trait;
use axum_core::extract::FromRef;
use serde::{Deserialize, Serialize};
use std::borrow::Borrow;
use std::str::FromStr;
use std::time::Duration;
use tracing::{debug, instrument};

pub struct PgEnqueuer;

#[async_trait]
impl Enqueuer for PgEnqueuer {
    type Error = crate::error::Error;

    #[instrument(skip_all)]
    async fn enqueue<W, S, Args, ArgsRef, E>(state: &S, args: ArgsRef) -> Result<(), Self::Error>
    where
        W: 'static + Worker<S, Args, Error = E>,
        S: Clone + Send + Sync + 'static,
        AppContext: FromRef<S>,
        Args: Send + Sync + Serialize + for<'de> Deserialize<'de>,
        ArgsRef: Send + Sync + Borrow<Args> + Serialize,
    {
        enqueue::enqueue::<W, _, _, _, _, _>(
            state,
            args,
            async |state, queue, job| -> RoadsterResult<()> {
                let context = AppContext::from_ref(state);
                let id = context.pgmq().send(queue, &job).await?;
                debug!(job.msg_id = id, "Job enqueued");
                Ok(())
            },
        )
        .await
    }

    #[instrument(skip_all)]
    async fn enqueue_delayed<W, S, Args, ArgsRef, E>(
        state: &S,
        args: ArgsRef,
        delay: Duration,
    ) -> Result<(), Self::Error>
    where
        W: 'static + Worker<S, Args, Error = E>,
        S: Clone + Send + Sync + 'static,
        AppContext: FromRef<S>,
        Args: Send + Sync + Serialize + for<'de> Deserialize<'de>,
        ArgsRef: Send + Sync + Borrow<Args> + Serialize,
    {
        enqueue::enqueue::<W, _, _, _, _, _>(
            state,
            args,
            async move |state, queue, job| -> RoadsterResult<()> {
                let context = AppContext::from_ref(state);
                let id = context
                    .pgmq()
                    .send_delay(queue, &job, delay.as_secs())
                    .await?;
                debug!(job.msg_id = id, job.delay = delay.as_secs(), "Job enqueued");
                Ok(())
            },
        )
        .await
    }

    #[instrument(skip_all)]
    async fn enqueue_batch<W, S, Args, ArgsRef, E>(
        state: &S,
        args: &[ArgsRef],
    ) -> Result<(), Self::Error>
    where
        W: 'static + Worker<S, Args, Error = E>,
        S: Clone + Send + Sync + 'static,
        AppContext: FromRef<S>,
        Args: Send + Sync + Serialize + for<'de> Deserialize<'de>,
        ArgsRef: Send + Sync + Borrow<Args> + Serialize,
    {
        enqueue::enqueue_batch::<W, _, _, _, _, _>(
            state,
            args,
            async |state, queue, jobs| -> RoadsterResult<()> {
                let context = AppContext::from_ref(state);
                let ids = context.pgmq().send_batch(queue, &jobs).await?;
                debug!(count = ids.len(), "Jobs enqueued");
                if debug_tracing_enabled(context.config()) {
                    ids.iter()
                        .for_each(|id| debug!(job.msg_id = id, "Job enqueued"));
                }
                Ok(())
            },
        )
        .await
    }

    #[instrument(skip_all)]
    async fn enqueue_batch_delayed<W, S, Args, ArgsRef, E>(
        state: &S,
        args: &[ArgsRef],
        delay: Duration,
    ) -> Result<(), Self::Error>
    where
        W: 'static + Worker<S, Args, Error = E>,
        S: Clone + Send + Sync + 'static,
        AppContext: FromRef<S>,
        Args: Send + Sync + Serialize + for<'de> Deserialize<'de>,
        ArgsRef: Send + Sync + Borrow<Args> + Serialize,
    {
        enqueue::enqueue_batch::<W, _, _, _, _, _>(
            state,
            args,
            async move |state, queue, jobs| -> RoadsterResult<()> {
                let context = AppContext::from_ref(state);
                let ids = context
                    .pgmq()
                    .send_batch_delay(queue, &jobs, delay.as_secs())
                    .await?;
                debug!(count = ids.len(), delay = delay.as_secs(), "Jobs enqueued");
                if debug_tracing_enabled(context.config()) {
                    ids.iter().for_each(|id| {
                        debug!(job.msg_id = id, job.delay = delay.as_secs(), "Job enqueued")
                    });
                }
                Ok(())
            },
        )
        .await
    }
}

fn debug_tracing_enabled(config: &AppConfig) -> bool {
    tracing::Level::from_str(&config.tracing.level)
        .map(|level| level >= tracing::Level::DEBUG)
        .unwrap_or_default()
}

#[cfg(test)]
mod tests {
    use crate::config::AppConfig;
    use crate::testing::snapshot::TestCase;
    use rstest::{fixture, rstest};

    #[fixture]
    fn case() -> TestCase {
        Default::default()
    }

    #[fixture]
    fn config() -> AppConfig {
        AppConfig::test(None).unwrap()
    }

    #[rstest]
    #[case("trace", true)]
    #[case("debug", true)]
    #[case("info", false)]
    #[case("warn", false)]
    #[case("error", false)]
    #[case("invalid", false)]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn debug_tracing_enabled(
        _case: TestCase,
        mut config: AppConfig,
        #[case] level: &str,
        #[case] expected: bool,
    ) {
        config.tracing.level = level.to_owned();
        let enabled = super::debug_tracing_enabled(&config);
        assert_eq!(enabled, expected);
    }
}
