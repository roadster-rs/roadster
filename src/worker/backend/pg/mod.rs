//! Background task queue service backed by Postgres using [pgmq](https://docs.rs/pgmq).

use crate::config::AppConfig;
use crate::worker::config::{BackoffStrategy, CompletedAction, PgWorkerConfig, RetryConfig};
use rand::Rng;
use std::cmp::min;
use std::sync::OnceLock;
use std::time::Duration;

pub mod enqueue;
pub(crate) mod periodic_job;
pub mod processor;

static DEFAULT_COMPLETED_ACTION: OnceLock<CompletedAction> = OnceLock::new();

/// Action to take if a job succeeds.
pub(crate) fn success_action<'a>(
    app_config: &'a AppConfig,
    worker_config: Option<&'a PgWorkerConfig>,
) -> &'a CompletedAction {
    worker_config
        .and_then(|config| config.success_action.as_ref())
        .or(app_config
            .service
            .worker
            .worker_config
            .pg
            .as_ref()
            .and_then(|config| config.success_action.as_ref()))
        .unwrap_or(DEFAULT_COMPLETED_ACTION.get_or_init(|| CompletedAction::Delete))
}

/// Action to take if a job fails.
pub(crate) fn failure_action<'a>(
    app_config: &'a AppConfig,
    worker_config: Option<&'a PgWorkerConfig>,
) -> &'a CompletedAction {
    worker_config
        .and_then(|config| config.failure_action.as_ref())
        .or(app_config
            .service
            .worker
            .worker_config
            .pg
            .as_ref()
            .and_then(|config| config.failure_action.as_ref()))
        .unwrap_or(DEFAULT_COMPLETED_ACTION.get_or_init(|| CompletedAction::Archive))
}

/// Calculate the retry delay based on the provided [`RetryConfig`]s and the number of attempts.
///
/// Return None if it should not retry.
pub(crate) fn retry_delay(
    app_config: &AppConfig,
    worker_retry_config: Option<&RetryConfig>,
    attempt_num: i32,
) -> Option<Duration> {
    let attempt_u32 = u32::try_from(attempt_num).ok()?;
    let attempt_num = usize::try_from(attempt_u32).unwrap_or(usize::MAX);

    let default_retry_config = app_config
        .service
        .worker
        .worker_config
        .retry_config
        .as_ref();

    let max_retries = worker_retry_config
        .and_then(|config| config.max_retries)
        .or(default_retry_config.and_then(|config| config.max_retries))
        .unwrap_or_default();

    if attempt_num > max_retries {
        return None;
    }

    let delay = worker_retry_config
        .and_then(|config| config.delay)
        .or(default_retry_config.and_then(|config| config.delay));
    let delay = match delay {
        Some(delay) => delay,
        None => {
            return None;
        }
    };

    let delay_offset = worker_retry_config
        .and_then(|config| config.delay_offset)
        .or(default_retry_config.and_then(|config| config.delay_offset));
    let delay = match delay_offset {
        Some(delay_offset) => delay.saturating_add(Duration::from_secs(
            rand::rng().random_range(0..delay_offset.as_secs()),
        )),
        None => delay,
    };

    let backoff_strategy = worker_retry_config
        .and_then(|config| config.backoff_strategy.as_ref())
        .or(default_retry_config.and_then(|config| config.backoff_strategy.as_ref()));
    let backoff_strategy = match backoff_strategy {
        Some(backoff_strategy) => backoff_strategy,
        None => {
            return Some(delay);
        }
    };

    let delay = match backoff_strategy {
        BackoffStrategy::Exponential => Duration::from_secs(
            delay
                .as_secs()
                .saturating_mul(2u64.saturating_pow(attempt_u32)),
        ),
        BackoffStrategy::Linear => delay.saturating_mul(attempt_u32),
        BackoffStrategy::None => delay,
    };

    let max_delay = worker_retry_config
        .and_then(|config| config.max_delay.as_ref())
        .or(default_retry_config.and_then(|config| config.max_delay.as_ref()));
    let delay = match max_delay {
        Some(max_delay) => min(*max_delay, delay),
        None => delay,
    };

    Some(delay)
}

// todo
#[cfg(test)]
mod tests {}
