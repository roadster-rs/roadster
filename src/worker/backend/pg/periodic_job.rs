use crate::worker::PeriodicArgsJson;
use crate::worker::job::JobMetadata;

// Todo: Not sure if this should be public yet.
#[serde_with::skip_serializing_none]
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize, bon::Builder, Eq, PartialEq)]
#[non_exhaustive]
pub(crate) struct PeriodicJob {
    pub(crate) metadata: JobMetadata,
    pub(crate) periodic: PeriodicConfig,
    pub(crate) args: serde_json::Value,
}

// Todo: Not sure if this should be public yet.
#[serde_with::skip_serializing_none]
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize, bon::Builder, Eq, PartialEq)]
#[non_exhaustive]
pub(crate) struct PeriodicConfig {
    pub(crate) hash: u64,
    pub(crate) schedule: cron::Schedule,
}

impl From<&PeriodicArgsJson> for PeriodicJob {
    fn from(value: &PeriodicArgsJson) -> Self {
        use std::hash::{DefaultHasher, Hash, Hasher};

        let mut hash = DefaultHasher::new();
        value.hash(&mut hash);
        let hash = hash.finish();

        PeriodicJob::builder()
            .args(value.args.clone())
            .metadata(
                JobMetadata::builder()
                    .worker_name(value.worker_name.clone())
                    .build(),
            )
            .periodic(
                PeriodicConfig::builder()
                    .hash(hash)
                    .schedule(value.schedule.clone())
                    .build(),
            )
            .build()
    }
}

#[cfg(test)]
#[cfg(any(feature = "worker-sidekiq", feature = "worker-pg"))]
mod tests {
    use crate::testing::snapshot::TestCase;
    use crate::worker::PeriodicArgsJson;
    use crate::worker::backend::pg::periodic_job::{PeriodicConfig, PeriodicJob};
    use crate::worker::job::JobMetadata;
    use cron::Schedule;
    use insta::{assert_json_snapshot, assert_snapshot};
    use rstest::{fixture, rstest};
    use std::hash::{DefaultHasher, Hasher};
    use std::str::FromStr;

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn periodic_job_ser_and_deser() {
        let job = super::PeriodicJob::builder()
            .args(serde_json::json!({"foo": "bar"}))
            .metadata(JobMetadata::builder().worker_name("foo").build())
            .periodic(
                PeriodicConfig::builder()
                    .hash(1234) // fake hash
                    .schedule(Schedule::from_str("* * * * * *").unwrap())
                    .build(),
            )
            .build();

        let ser = serde_json::to_value(&job).unwrap();

        let job_deser: PeriodicJob = serde_json::from_value(ser).unwrap();

        assert_eq!(job, job_deser);
    }

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn periodic_job_serde() {
        let _case = TestCase::new();

        let job = super::PeriodicJob::builder()
            .args(serde_json::json!({"foo": "bar"}))
            .metadata(JobMetadata::builder().worker_name("foo").build())
            .periodic(
                PeriodicConfig::builder()
                    .hash(1234) // fake hash
                    .schedule(Schedule::from_str("* * * * * *").unwrap())
                    .build(),
            )
            .build();

        assert_json_snapshot!(job);
    }

    #[fixture]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn periodic_args_json() -> PeriodicArgsJson {
        PeriodicArgsJson::builder()
            .worker_name("a".to_string())
            .schedule(Schedule::from_str("* * * * * *").unwrap())
            .args(serde_json::json!({"foo": "bar"}))
            .build()
    }

    #[rstest]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn periodic_args_json_hash(periodic_args_json: PeriodicArgsJson) {
        let mut hasher = DefaultHasher::new();
        crate::worker::job::periodic_hash(
            &mut hasher,
            &periodic_args_json.worker_name,
            &periodic_args_json.schedule,
            &periodic_args_json.args,
        );
        assert_snapshot!(hasher.finish());
    }

    #[rstest]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn job_from_periodic_args_hash(periodic_args_json: PeriodicArgsJson) {
        let job = PeriodicJob::from(&periodic_args_json);
        let mut hasher = DefaultHasher::new();
        crate::worker::job::periodic_hash(
            &mut hasher,
            &job.metadata.worker_name,
            &job.periodic.schedule,
            &job.args,
        );
        assert_eq!(hasher.finish(), job.periodic.hash);
    }
}
