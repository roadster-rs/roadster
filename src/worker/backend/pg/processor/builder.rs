use crate::app::context::AppContext;
use crate::error::RoadsterResult;
use crate::worker::Worker;
use crate::worker::backend::pg::processor::{
    PgProcessor, PgProcessorError, ProcessorInner, WorkerWrapper,
};
use axum_core::extract::FromRef;
use cron::Schedule;
use serde::{Deserialize, Serialize};
use std::cmp::Ordering;
use tracing::{error, info};

#[non_exhaustive]
pub struct PgProcessorBuilder<S>
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
{
    pub(crate) inner: ProcessorInner<S>,
}

#[derive(bon::Builder)]
#[non_exhaustive]
pub struct PeriodicArgs<Args>
where
    Args: Send + Sync + Serialize + for<'de> Deserialize<'de>,
{
    args: Args,
    schedule: Schedule,
}

#[derive(Clone, bon::Builder, Eq, PartialEq)]
#[non_exhaustive]
pub(crate) struct PeriodicArgsJson {
    pub(crate) args: serde_json::Value,
    pub(crate) worker_name: String,
    pub(crate) schedule: Schedule,
}

impl Ord for PeriodicArgsJson {
    fn cmp(&self, other: &Self) -> Ordering {
        self.worker_name
            .cmp(&other.worker_name)
            .then(self.schedule.to_string().cmp(&other.schedule.to_string()))
            .then(
                serde_json::to_string(&self.args)
                    .unwrap_or_default()
                    .cmp(&serde_json::to_string(&other.args).unwrap_or_default()),
            )
    }
}

impl PartialOrd for PeriodicArgsJson {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl<S> PgProcessorBuilder<S>
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
{
    pub(crate) fn new(state: &S) -> Self {
        Self {
            inner: ProcessorInner {
                state: state.clone(),
                queues: Default::default(),
                workers: Default::default(),
                periodic_workers: Default::default(),
            },
        }
    }

    pub fn build(self) -> PgProcessor<S> {
        PgProcessor::new(self.inner)
    }

    pub fn register<W, Args, E>(mut self, worker: W) -> RoadsterResult<Self>
    where
        W: 'static + Worker<S, Args, Error = E>,
        Args: Send + Sync + Serialize + for<'de> Deserialize<'de>,
        // Todo: without this `'static`, we're getting an internal compiler error
        E: 'static + std::error::Error + Send + Sync,
    {
        let name = W::name();
        info!(name, "Registering PG worker");

        self.register_internal(worker, name, false)?;

        Ok(self)
    }

    pub fn register_periodic<W, Args, E>(
        mut self,
        worker: W,
        periodic_args: PeriodicArgs<Args>,
    ) -> RoadsterResult<Self>
    where
        W: 'static + Worker<S, Args, Error = E>,
        Args: Send + Sync + Serialize + for<'de> Deserialize<'de>,
        // Todo: without this `'static`, we're getting an internal compiler error
        E: 'static + std::error::Error + Send + Sync,
    {
        let name = W::name();
        info!(name, "Registering periodic PG worker");

        self.register_internal(worker, name.clone(), true)?;

        let periodic_args = PeriodicArgsJson::builder()
            .args(serde_json::to_value(periodic_args.args)?)
            .worker_name(name.clone())
            .schedule(periodic_args.schedule)
            .build();

        if let Some(replaced) = self.inner.periodic_workers.replace(periodic_args) {
            return Err(PgProcessorError::AlreadyRegisteredPeriodic(
                replaced.worker_name,
                replaced.schedule,
                replaced.args,
            )
            .into());
        }

        Ok(self)
    }

    fn register_internal<W, Args, E>(
        &mut self,
        worker: W,
        name: String,
        skip_duplicate: bool,
    ) -> RoadsterResult<()>
    where
        W: 'static + Worker<S, Args, Error = E>,
        Args: Send + Sync + Serialize + for<'de> Deserialize<'de>,
        // Todo: without this `'static`, we're getting an internal compiler error
        E: 'static + std::error::Error + Send + Sync,
    {
        let context = AppContext::from_ref(&self.inner.state);
        let enqueue_config = &context.config().service.worker.enqueue_config;
        let worker_enqueue_config = W::enqueue_config(&self.inner.state);

        let queue = worker_enqueue_config
            .queue
            .as_ref()
            .or(enqueue_config.queue.as_ref());
        let queue = if let Some(queue) = queue {
            queue
        } else {
            error!(
                worker_name = W::name(),
                "Unable to register worker, no queue configured"
            );
            return Err(PgProcessorError::NoQueue(W::name()).into());
        };
        self.inner.queues.insert(queue.to_owned());

        if self
            .inner
            .workers
            .insert(
                name.clone(),
                WorkerWrapper::new(&self.inner.state, worker, worker_enqueue_config)?,
            )
            .is_some()
            && !skip_duplicate
        {
            return Err(PgProcessorError::AlreadyRegistered(name).into());
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    mod periodic_args {
        use crate::worker::backend::pg::processor::builder::PeriodicArgsJson;
        use crate::worker::job::Job;
        use cron::Schedule;
        use insta::assert_json_snapshot;
        use rstest::{fixture, rstest};
        use std::str::FromStr;

        #[fixture]
        #[cfg_attr(coverage_nightly, coverage(off))]
        fn periodic_args_json() -> PeriodicArgsJson {
            PeriodicArgsJson::builder()
                .worker_name("a".to_string())
                .schedule(Schedule::from_str("* * * * * *").unwrap())
                .args(serde_json::json!({"foo": "bar"}))
                .build()
        }

        #[rstest]
        #[cfg_attr(coverage_nightly, coverage(off))]
        fn periodic_args_json_ord_name(periodic_args_json: PeriodicArgsJson) {
            let mut b = periodic_args_json.clone();
            b.worker_name = "b".to_string();
            assert!(periodic_args_json < b);
        }

        #[rstest]
        #[cfg_attr(coverage_nightly, coverage(off))]
        fn periodic_args_json_ord_schedule(periodic_args_json: PeriodicArgsJson) {
            let mut b = periodic_args_json.clone();
            b.schedule = Schedule::from_str("*/10 * * * * *").unwrap();
            assert!(periodic_args_json < b);
        }

        #[rstest]
        #[cfg_attr(coverage_nightly, coverage(off))]
        fn periodic_args_json_ord_args(periodic_args_json: PeriodicArgsJson) {
            let mut b = periodic_args_json.clone();
            b.args = serde_json::json!({"foo": "baz"});
            assert!(periodic_args_json < b);
        }

        #[rstest]
        #[cfg_attr(coverage_nightly, coverage(off))]
        fn job_from_periodic_args(periodic_args_json: PeriodicArgsJson) {
            let job = Job::from(&periodic_args_json);
            assert_json_snapshot!(job);
        }

        #[rstest]
        #[cfg_attr(coverage_nightly, coverage(off))]
        fn job_from_periodic_args_hash(periodic_args_json: PeriodicArgsJson) {
            let job = Job::from(&periodic_args_json);
            let hash = crate::worker::job::periodic_hash(
                &job.metadata.worker_name,
                &job.metadata.periodic.as_ref().unwrap().schedule,
                &job.args,
            );
            assert_eq!(hash, job.metadata.periodic.unwrap().hash);
        }
    }
}
