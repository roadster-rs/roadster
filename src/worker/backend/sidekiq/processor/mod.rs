use crate::app::context::AppContext;
use crate::config::service::worker::StaleCleanUpBehavior;
use crate::error::RoadsterResult;
use crate::util::redis::RedisCommands;
use crate::worker::backend::sidekiq::processor::builder::SidekiqProcessorBuilder;
use crate::worker::{
    PeriodicArgsJson, RegisterSidekiqFn, RegisterSidekiqPeriodicFn, WorkerWrapper,
};
use axum_core::extract::FromRef;
use cron::Schedule;
use itertools::Itertools;
use sidekiq::periodic;
use std::collections::{BTreeMap, BTreeSet, HashSet};
use std::sync::{Arc, Mutex};
use thiserror::Error;
use tokio::task::JoinSet;
use tokio_util::sync::CancellationToken;
use tracing::{error, info, warn};

pub mod builder;

const PERIODIC_KEY: &str = "periodic";

#[derive(Debug, Error)]
#[non_exhaustive]
pub enum SidekiqProcessorError {
    /// The provided [`Worker`] was already registered. Contains the [`Worker::name`]
    /// of the provided worker.
    #[error("The provided `Worker` name was already registered: `{0}`")]
    AlreadyRegistered(String),

    /// A [`Worker`] was previously registered that has the same name but is a different type.
    #[error("The provided `Worker` name was already registered for a different type: `{0}`")]
    AlreadyRegisteredWithDifferentType(String),

    /// The provided [`Worker`] was already registered. Contains the [`Worker::name`]
    /// of the provided worker.
    #[error(
        "The provided periodic worker job was already registered. Worker: `{0}`, schedule: `{1}`, args: `{2}`"
    )]
    AlreadyRegisteredPeriodic(String, Schedule, serde_json::Value),

    #[error("No queue configured for worker `{0}`.")]
    NoQueue(String),

    #[error(transparent)]
    Other(#[from] Box<dyn std::error::Error + Send + Sync>),
}

#[derive(Clone)]
#[non_exhaustive]
pub struct SidekiqProcessor<S>
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
{
    inner: Arc<SidekiqProcessorInner<S>>,
}

pub(crate) struct WorkerData<S>
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
{
    pub(crate) worker_wrapper: WorkerWrapper<S>,
    pub(crate) register_sidekiq_fn: RegisterSidekiqFn<S>,
    pub(crate) register_sidekiq_periodic_fn: RegisterSidekiqPeriodicFn<S>,
}

#[non_exhaustive]
pub(crate) struct SidekiqProcessorInner<S>
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
{
    state: S,
    // Todo: we may need to register directly on the processor instead of waiting to register
    //  until later, depending on if `RoadsterWorker` needs the `W` type param.
    // todo: store a closure to register the worker in order to keep the type?
    processor: Mutex<Option<::sidekiq::Processor>>,
    // queues: BTreeSet<String>,
    workers: BTreeMap<String, WorkerData<S>>,
    periodic_workers: HashSet<PeriodicArgsJson>,
}

impl<S> SidekiqProcessor<S>
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
{
    pub(crate) fn new(inner: SidekiqProcessorInner<S>) -> Self {
        Self {
            inner: Arc::new(inner),
        }
    }

    pub fn builder(state: &S) -> SidekiqProcessorBuilder<S> {
        SidekiqProcessorBuilder::new(state)
    }

    pub async fn before_run(&self, state: &S) -> RoadsterResult<()> {
        self.initialize_periodic(state).await?;

        Ok(())
    }

    /// Initialize the periodic queue tables and enqueue the periodic jobs in the queue.
    async fn initialize_periodic(&self, state: &S) -> RoadsterResult<()> {
        let context = AppContext::from_ref(state);

        let periodic_config = &context
            .config()
            .service
            .worker
            .sidekiq
            .custom
            .custom
            .periodic;

        match periodic_config.stale_cleanup {
            StaleCleanUpBehavior::Manual => {}
            StaleCleanUpBehavior::AutoCleanAll => {
                periodic::destroy_all(context.redis_enqueue().inner.clone()).await?;
                info!("Deleted all previously registered periodic jobs");
            }
            StaleCleanUpBehavior::AutoCleanStale => {
                let mut conn = context.redis_enqueue().get().await?;
                // todo: replace `default` with actual registered periodic jobs
                remove_stale_periodic_jobs(&mut conn, &context, &Default::default()).await?;
            }
        };

        let mut processor = self.inner.processor.lock()?;

        let processor = if let Some(processor) = processor.as_mut() {
            processor
        } else {
            return Err(todo!());
        };

        for periodic_args in self.inner.periodic_workers.iter() {
            if let Some(worker_data) = self.inner.workers.get(&periodic_args.worker_name) {
                (worker_data.register_sidekiq_periodic_fn)(
                    &self.inner.state,
                    processor,
                    worker_data.worker_wrapper.clone(),
                    periodic_args.clone(),
                )
                .await?;
            } else {
                return Err(todo!());
            }
        }

        Ok(())
    }

    pub async fn run(self, _state: &S, cancellation_token: CancellationToken) {
        let processor = {
            match self.inner.processor.lock() {
                Ok(processor) => processor.clone(),
                Err(err) => {
                    error!("Unable to lock ::sidekiq::Processor: {err}");
                    cancellation_token.cancel();
                    return;
                }
            }
        };

        let processor = match processor.clone() {
            Some(processor) => processor,
            None => {
                warn!("No ::sidekiq::Processor available.");
                return;
            }
        };
        let sidekiq_cancel_token = processor.get_cancellation_token();

        let mut join_set = JoinSet::new();
        let token = cancellation_token.clone();
        join_set.spawn(Box::pin(async move {
            token.cancelled().await;
        }));
        let token = sidekiq_cancel_token.clone();
        join_set.spawn(Box::pin(async move {
            token.cancelled().await;
        }));
        join_set.spawn(processor.run());

        while let Some(result) = join_set.join_next().await {
            // Once any of the tasks finish, cancel all the cancellation tokens to ensure
            // the processor and the app shut down gracefully.
            cancellation_token.cancel();
            sidekiq_cancel_token.cancel();
            if let Err(join_err) = result {
                error!(
                    "An error occurred when trying to join on one of the app's tasks. Error: {join_err}"
                );
            }
        }
    }
}

/// Compares the list of periodic jobs that were registered by the app during app startup with
/// the list of periodic jobs in Redis, and removes any that exist in Redis but weren't
/// registered during start up.
///
/// The jobs are only removed if the [worker.sidekiq.periodic.stale-cleanup][crate::config::worker::Periodic]
/// config is set to [auto-clean-stale][StaleCleanUpBehavior::AutoCleanStale].
///
/// This is run after all the app's periodic jobs have been registered.
async fn remove_stale_periodic_jobs<C: RedisCommands>(
    conn: &mut C,
    context: &AppContext,
    registered_periodic_workers: &HashSet<String>,
) -> RoadsterResult<()> {
    let stale_jobs = conn
        .zrange(PERIODIC_KEY.to_string(), 0, -1)
        .await?
        .into_iter()
        .filter(|job| !registered_periodic_workers.contains(job))
        .collect_vec();

    if stale_jobs.is_empty() {
        info!("No stale periodic jobs found");
        return Ok(());
    }

    if context
        .config()
        .service
        .worker
        .sidekiq
        .custom
        .custom
        .periodic
        .stale_cleanup
        == StaleCleanUpBehavior::AutoCleanStale
    {
        info!(count = stale_jobs.len(), "Removing stale periodic jobs",);
        conn.zrem(PERIODIC_KEY.to_string(), stale_jobs.clone())
            .await?;
    } else {
        warn!(count = stale_jobs.len(), "Found stale periodic jobs");
    }

    Ok(())
}
