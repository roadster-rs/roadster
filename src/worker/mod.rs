pub mod app_worker;
pub mod registry;

use crate::app::App;

use crate::worker::app_worker::AppWorkerConfig;
use app_worker::AppWorker;
use async_trait::async_trait;
use serde::Serialize;

use sidekiq::{RedisPool, Worker, WorkerOpts};
use std::marker::PhantomData;
use std::sync::Arc;
use std::time::Duration;
use tracing::{error, instrument};

/// Worker used by Roadster to wrap the consuming app's workers to add additional behavior. For
/// example, [RoadsterWorker] is by default configured to automatically abort the app's worker
/// when it exceeds a certain timeout.
pub struct RoadsterWorker<A, Args, W>
where
    A: App,
    Args: Send + Sync + Serialize + 'static,
    W: AppWorker<A, Args>,
{
    inner: W,
    inner_config: AppWorkerConfig,
    _args: PhantomData<Args>,
    _app: PhantomData<A>,
}

impl<A, Args, W> RoadsterWorker<A, Args, W>
where
    A: App,
    Args: Send + Sync + Serialize,
    W: AppWorker<A, Args>,
{
    pub(crate) fn new(inner: W, state: Arc<A::State>) -> Self {
        let config = inner.config(&state);
        Self {
            inner,
            inner_config: config,
            _args: PhantomData,
            _app: PhantomData,
        }
    }
}

#[async_trait]
impl<A, Args, W> Worker<Args> for RoadsterWorker<A, Args, W>
where
    A: App,
    Args: Send + Sync + Serialize,
    W: AppWorker<A, Args>,
{
    fn disable_argument_coercion(&self) -> bool {
        self.inner_config.disable_argument_coercion
    }

    fn opts() -> WorkerOpts<Args, Self>
    where
        Self: Sized,
    {
        // This method not implemented because `RoadsterWorker` should not be enqueued directly,
        // and this method is only used when enqueuing. Instead, Sidekiq.rs will use the
        // `W::opts` implementation directly.
        unimplemented!()
    }

    fn max_retries(&self) -> usize {
        self.inner_config.max_retries
    }

    fn class_name() -> String
    where
        Self: Sized,
    {
        // This method is implemented because it's used both when registering the worker, and
        // when enqueuing a job. We forward the implementation to `W::classname` because that's
        // what Sidekiq.rs uses specifically. If we attempt to override this, our impl will be used
        // when registering the worker, but not when enqueuing a job, so the worker will not pick
        // up the jobs.
        W::class_name()
    }

    async fn perform_async(_redis: &RedisPool, _args: Args) -> sidekiq::Result<()>
    where
        Self: Sized,
        Args: Send + Sync + Serialize + 'static,
    {
        // This method not implemented because `RoadsterWorker` should not be enqueued directly.
        unimplemented!()
    }

    async fn perform_in(_redis: &RedisPool, _duration: Duration, _args: Args) -> sidekiq::Result<()>
    where
        Self: Sized,
        Args: Send + Sync + Serialize + 'static,
    {
        // This method not implemented because `RoadsterWorker` should not be enqueued directly.
        unimplemented!()
    }

    #[instrument(skip_all)]
    async fn perform(&self, args: Args) -> sidekiq::Result<()> {
        let inner = self.inner.perform(args);

        if self.inner_config.timeout {
            tokio::time::timeout(self.inner_config.max_duration, inner)
                .await
                .map_err(|err| {
                    error!(
                        worker = %W::class_name(),
                        max_duration = %self.inner_config.max_duration.as_secs(),
                        %err,
                        "Worker timed out"
                    );
                    sidekiq::Error::Any(Box::new(err))
                })?
        } else {
            inner.await
        }
    }
}
