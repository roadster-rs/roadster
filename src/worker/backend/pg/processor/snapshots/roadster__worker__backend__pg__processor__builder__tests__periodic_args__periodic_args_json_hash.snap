---
source: src/worker/backend/pg/processor/builder.rs
expression: hasher.finish()
---
11189630487790136136
